File: mypython/dataset.py
"""Character-level tokenizer + JSONL corpus loader.

We keep tokenization deliberately simple (one token == one character) so the
model trains in seconds-to-minutes on a laptop. The vocabulary is built from the
training text itself and stored inside the checkpoint, so generation uses the
exact same mapping.
"""
import json


class CharTokenizer:
    def __init__(self, stoi: dict, itos: dict):
        self.stoi = stoi                 # char -> id
        self.itos = itos                 # id   -> char
        self.vocab_size = len(stoi)

    @classmethod
    def from_text(cls, text: str) -> "CharTokenizer":
        chars = sorted(set(text))
        stoi = {ch: i for i, ch in enumerate(chars)}
        itos = {i: ch for i, ch in enumerate(chars)}
        return cls(stoi, itos)

    def encode(self, s: str) -> list[int]:
        return [self.stoi[c] for c in s if c in self.stoi]

    def decode(self, ids: list[int]) -> str:
        return "".join(self.itos[i] for i in ids)


def load_jsonl_text(path: str) -> str:
    """Read a JSONL file of {"text": ...} records and join them into one string."""
    texts = []
    with open(path, encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            texts.append(json.loads(line)["text"])
    return "\n".join(texts)

File: myrust/src/dataset.rs
//! Character-level tokenizer + JSONL corpus loader (Python: dataset.py).
//!
//! One token == one character. The vocabulary is built from the training text
//! itself (sorted unique chars) and stored inside the checkpoint, so generation
//! uses the exact same mapping.
use std::collections::HashMap;
use std::fs;

use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CharTokenizer {
    /// The vocabulary, in id order: `vocab[id] == char`. This single ordered
    /// list captures both the id->char (itos) and char->id (stoi) mappings.
    vocab: Vec<char>,
    #[serde(skip)]
    stoi: HashMap<char, usize>,
}

impl CharTokenizer {
    fn from_vocab(vocab: Vec<char>) -> Self {
        let stoi = vocab.iter().enumerate().map(|(i, &c)| (c, i)).collect();
        Self { vocab, stoi }
    }

    /// Build a tokenizer from text: the vocabulary is the sorted set of chars,
    /// matching Python's `sorted(set(text))`.
    pub fn from_text(text: &str) -> Self {
        let mut chars: Vec<char> = text.chars().collect::<std::collections::BTreeSet<_>>()
            .into_iter()
            .collect();
        chars.sort_unstable(); // BTreeSet is already sorted; explicit for clarity
        Self::from_vocab(chars)
    }

    /// Rebuild a tokenizer from a saved vocabulary string (chars in id order).
    pub fn from_vocab_string(vocab: &str) -> Self {
        Self::from_vocab(vocab.chars().collect())
    }

    /// The vocabulary as a single string, chars in id order — what we persist.
    pub fn vocab_string(&self) -> String {
        self.vocab.iter().collect()
    }

    pub fn vocab_size(&self) -> usize {
        self.vocab.len()
    }

    /// Encode text to token ids, silently dropping out-of-vocabulary chars
    /// (matches Python's `if c in self.stoi`).
    pub fn encode(&self, s: &str) -> Vec<i64> {
        s.chars()
            .filter_map(|c| self.stoi.get(&c).map(|&i| i as i64))
            .collect()
    }

    /// Decode token ids back to text.
    pub fn decode(&self, ids: &[i64]) -> String {
        ids.iter()
            .filter_map(|&i| self.vocab.get(i as usize))
            .collect()
    }
}

/// Read a JSONL file of `{"text": ...}` records and join them into one string
/// (Python: load_jsonl_text).
pub fn load_jsonl_text(path: &str) -> std::io::Result<String> {
    let raw = fs::read_to_string(path)?;
    let mut texts = Vec::new();
    for line in raw.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let v: serde_json::Value = serde_json::from_str(line)?;
        if let Some(t) = v.get("text").and_then(|t| t.as_str()) {
            texts.push(t.to_string());
        }
    }
    Ok(texts.join("\n"))
}
